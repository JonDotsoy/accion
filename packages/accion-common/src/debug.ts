import { Style, styleText } from "./style-text.js";

const schemaColorsRandom = [
    "red",
    "green",
    "yellow",
    "blue",
    "magenta",
    "cyan",
].sort(() => (Math.random() > 0.5 ? 1 : -1)) as Style[];

const indexLogRef = { current: 0 };

const minuteFormat = new Intl.NumberFormat(undefined, {
    style: "unit",
    unit: "minute",
    unitDisplay: "narrow",
});
const secondFormat = new Intl.NumberFormat(undefined, {
    style: "unit",
    unit: "second",
    unitDisplay: "narrow",
});
const millisecondFormat = new Intl.NumberFormat(undefined, {
    style: "unit",
    unit: "millisecond",
    unitDisplay: "narrow",
});

const durationFormat = (durationMillisecond: number) => {
    if (durationMillisecond < 1000)
        return millisecondFormat.format(durationMillisecond);
    if (durationMillisecond < 60000)
        return secondFormat.format(durationMillisecond / 1000);
    return minuteFormat.format(durationMillisecond / 60000);
};

export const makeIsEnabled = (debugExpression?: string) => {
    const rules =
        debugExpression
            ?.split(",")
            .filter((chunk) => chunk.length)
            .map(
                (chunk) =>
                    `^${chunk.replace(/\*|\W/g, (char) => (char === "*" ? ".*?" : `\\${char}`))}$`,
            )
            .map((chunk) => new RegExp(chunk)) ?? [];

    return (namespace: string) => {
        if (!rules.length) return false;
        return rules.some((exp) => exp.test(namespace));
    };
};

const globalIsEnabled = makeIsEnabled(process.env.DEBUG ?? "");

export const debug = (namespace: string) => {
    let lastCall = Date.now();
    const indexLog = indexLogRef.current++;
    const schemaColor: Style =
        schemaColorsRandom[indexLog % schemaColorsRandom.length];
    const log = (message: string) => {
        const l = Date.now() - lastCall;
        lastCall = Date.now();
        if (log.enabled) {
            const latency = `${l >= 0 ? "+" : "-"}${durationFormat(l)}`;
            console.log(
                `${styleText(schemaColor, namespace)} ${message} ${styleText(schemaColor, latency)}`,
            );
        }
    };
    log.namespace = namespace;
    log.enabled = globalIsEnabled(namespace);
    log.extend = (childNamespace: string = "") =>
        debug(`${namespace}${childNamespace}`);
    return log;
};
