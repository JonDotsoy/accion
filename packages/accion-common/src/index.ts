export { debug } from "./debug";
export { exec } from "./exec";
export { promiseWithResolvers } from "./promise-with-resolvers";
