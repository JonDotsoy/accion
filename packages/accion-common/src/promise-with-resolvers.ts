const noop: (...args: any[]) => void = () => {};

export const promiseWithResolvers = <T extends any>() => {
    let resolve: (value: T) => void = noop;
    let reject: (reason: any) => void = noop;

    const promise = new Promise<T>((innerResolve, innerReject) => {
        resolve = innerResolve;
        reject = innerReject;
    });

    if (resolve === noop) throw new Error(`Cannot load resolve function`);
    if (reject === noop) throw new Error(`Cannot load reject function`);

    return {
        promise,
        resolve,
        reject,
    };
};
