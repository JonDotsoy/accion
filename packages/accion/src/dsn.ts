import { nextTick } from "process";
import { Event } from "./interfaces/event";
import type { Manager } from "./manager";
import { atom } from "nanostores";

export class Closing {
  waitings = new Set<Promise<unknown>>();
}

export abstract class DSN {
  static create(
    manager: Manager,
    source: DSN | string,
    options?: { signal?: AbortSignal },
  ) {
    if (source instanceof DSN) return source;
    if (typeof source === "string")
      return new FetchDSN(manager, source, { signal: options?.signal });
    throw new Error(`Invalid dsn source`);
  }
}

type FetchDSNOptions = {
  signal?: AbortSignal;
  limitBytesPerRequest?: number;
};

export class FetchDSN extends DSN {
  eventsLayer = new Set<Event>();
  eventsLayerIds = new Map<Event, string>();
  limitBytesPerRequest: number;
  requestLoopStack = atom(new Set<Request>());
  signal?: AbortSignal;
  preventStopped = false;

  constructor(
    readonly manager: Manager,
    readonly uri: string,
    readonly options?: FetchDSNOptions,
  ) {
    super();
    this.signal = options?.signal;
    this.limitBytesPerRequest = options?.limitBytesPerRequest ?? 8000;
    if (this.signal?.aborted) {
      throw new Error(`Signal already aborted`);
    }
    const unsubscribeManager = manager.subscribeChange((event) => {
      this.eventsLayer.add(event);
    });

    this.signal?.addEventListener("abort", () => {
      const reason = this.signal?.reason;
      unsubscribeManager();
      if (reason instanceof Closing) {
        this.preventStopped = true;
        const { promise, resolve } = Promise.withResolvers<void>();
        this.requestLoopStack.subscribe(() => {
          if (
            this.eventsLayer.size === 0 &&
            this.requestLoopStack.get().size === 0
          ) {
            resolve();
            this.preventStopped = false;
            this.requestLoopStack.off();
          }
        });
        reason.waitings.add(promise);
      }
    });
    this.startSyncLoop();
  }

  private isStopped() {
    if (this.preventStopped) return false;
    return this.signal?.aborted ?? false;
  }

  readEventLayer() {
    const bytes: number[] = [];
    for (const event of this.eventsLayer) {
      const bytesLine = new TextEncoder().encode(`${JSON.stringify(event)}\n`);
      if (bytes.length + bytesLine.length > this.limitBytesPerRequest) {
        break;
      }
      bytes.push(...bytesLine);
      this.eventsLayer.delete(event);
    }

    if (bytes.length === 0) return;

    const headers = new Headers();
    headers.set("Content-Type", "application/jsonl; charset=utf-8");
    headers.set("Date", new Date().toUTCString());
    const request = new Request({
      method: "POST",
      url: this.uri,
      headers,
      body: new Uint8Array(bytes),
      signal: this.signal,
    });

    const nextRequestLoopStack = new Set(this.requestLoopStack.get());
    nextRequestLoopStack.add(request);
    this.requestLoopStack.set(nextRequestLoopStack);
  }

  startSyncLoop() {
    const loadProcess = async () => {
      if (this.isStopped()) return { fetches: 0 };

      let fetches = 0;
      for (const request of this.requestLoopStack.get()) {
        fetches += 1;
        await fetch(request);
        const nextRequestLoopStack = new Set(this.requestLoopStack.get());
        nextRequestLoopStack.delete(request);
        this.requestLoopStack.set(nextRequestLoopStack);
      }
      return { fetches };
    };

    const loopAsync = async () => {
      while (!this.isStopped()) {
        const { promise, resolve } = Promise.withResolvers<void>();
        setTimeout(() => resolve(), 50);
        this.readEventLayer();
        await loadProcess();
        await promise;
      }
    };
    loopAsync();
  }
}
