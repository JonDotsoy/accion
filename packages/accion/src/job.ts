export type { JobContext, JobReturn, JobArgs, JobCb } from "./interfaces/job.ts"
import type { JobContext, JobReturn, JobArgs, JobCb } from "./interfaces/job.ts"

const isCb = (value: unknown): value is (() => JobCb<any>) => typeof value === 'function'

export const job = <R extends JobReturn, I extends Record<string, JobContext<any>>>(
    ...args: JobArgs<R, I>
): JobContext<R> => {
    let id: null | string = null
    let description: null | string = null
    let cb: null | (() => JobCb<R>) = null
    let input: null | I = null

    for (const arg of args) {
        if (typeof arg === 'string') id = arg
        if (isCb(arg)) cb = arg
        if (typeof arg === 'object' && arg !== null) {
            id = id ?? arg.id ?? null
            description = arg.description ?? null
            input = arg.input ?? null
        }
    }

    if (id === null) throw new Error(`Missing id`)
    if (cb === null) throw new Error(`Missing cb function`)

    return {
        id,
        description,
        input,
        cb,
    }
}
