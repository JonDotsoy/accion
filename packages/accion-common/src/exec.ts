import { debug } from "./debug.js";
import * as child_process from "child_process";
import { promiseWithResolvers } from "./promise-with-resolvers.js";

type Options = {
    returnOutput?: boolean;
    silence?: boolean;
    verbose?: boolean;
    shell?: string;
};

export const exec = async (
    cmd: string,
    ...args: [args?: string[], options?: Options] | [options?: Options]
) => {
    const log = debug("exec");
    let argOptions: Options = {};
    let cmdArgs: null | string[] = null;

    for (const arg of args) {
        if (Array.isArray(arg)) {
            cmdArgs = arg;
            continue;
        }
        if (typeof arg === "object" && arg !== null) {
            argOptions = arg;
            continue;
        }
    }

    const omitReturnOutput = argOptions?.returnOutput === false;
    const verbose = argOptions?.verbose === true || log.enabled;
    const shell = argOptions?.shell ?? "sh";
    const printStdoutAndStderr = verbose
        ? true
        : argOptions.silence
          ? false
          : true;

    if (verbose) log.enabled = true;

    const { resolve, reject, promise } = promiseWithResolvers();

    log(`Run ${cmd}${cmdArgs ? ` ${cmdArgs.join(" ")}` : ""}`);

    const matrixCmd = cmdArgs
        ? { cmd, cmdArgs }
        : { cmd: shell, cmdArgs: ["-c", cmd] };

    const subprocess = child_process.spawn(
        matrixCmd.cmd,
        matrixCmd.cmdArgs ?? [],
        {
            stdio: "pipe",
            detached: false,
            shell: false,
        },
    );

    subprocess.addListener("error", (error) => reject(error));
    subprocess.addListener("exit", (code) => resolve(code));

    let outputBuff = new Uint8Array([]);

    subprocess.stdout.addListener("data", (data) => {
        if (printStdoutAndStderr) process.stdout.write(data);
        if (!omitReturnOutput) {
            outputBuff = new Uint8Array([...outputBuff, ...data]);
        }
    });

    subprocess.stderr.addListener("data", (data) => {
        if (printStdoutAndStderr) process.stdout.write(data);
    });

    await promise;

    const text = () => new TextDecoder().decode(outputBuff);
    const json = () => JSON.parse(text());

    if (subprocess.exitCode !== 0) throw new Error(`Exists status`);

    return { exitCode: subprocess.exitCode, text, json };
};
