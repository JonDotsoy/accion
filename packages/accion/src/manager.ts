export type { JobContext } from "./interfaces/job.ts";
export type { StorageService } from "./interfaces/storage-service.ts";
import { nextTick } from "process";
import { DSN, Closing } from "./dsn.ts";
import { State } from "./dto/state.ts";
import { Event } from "./interfaces/event.ts";
import type { JobContext } from "./interfaces/job.ts";
import type { StorageService } from "./interfaces/storage-service.ts";

type JobState = {
  state: State;
};

type RawJson = {
  indexes: string[];
  relations: [string, string][];
  jobStates: Record<string, JobState>;
};

namespace RawJsonSchemaValidator {
  type Ctx = {
    path: string[];
  };

  const wrapVal = <T>(ctx: Ctx, cb: (value: T) => boolean) => {
    return (value: T) => {
      const res = cb(value);

      return res;
    };
  };
  const isObject = (value: unknown): value is Record<string, unknown> =>
    typeof value === "object" && value !== null;
  const isString = (value: unknown): value is string =>
    typeof value === "string";

  export const isRawJson = (
    value: unknown,
    ctx: Ctx = { path: [] },
  ): value is RawJson =>
    isObject(value) &&
    Array.isArray(value.indexes) &&
    value.indexes.every(isString) &&
    Array.isArray(value.relations) &&
    value.relations.every(([k, p]) => isString(k) && isString(p)) &&
    isObject(value.jobStates) &&
    Object.entries(value.jobStates).every(
      ([k, jobState]) =>
        isString(k) && isObject(jobState) && isString(jobState.state),
    );
}

const isRawJson = RawJsonSchemaValidator.isRawJson;

const createLocalStorageService = (): StorageService =>
  new Map<string, unknown>() as unknown as StorageService;

export class Manager {
  private relations: [JobContext<any>, JobContext<any>][] = [];
  private jobStates = new Map<JobContext<any>, JobState>();
  private jobIndexes: Map<string, JobContext<any>>;
  private subs = new Set<(event: Event) => void>();
  private storageService: StorageService = createLocalStorageService();

  private constructor(
    readonly sid: string,
    readonly did: string | undefined,
    private indexes: Map<JobContext<any>, string>,
    readonly jobs: Set<JobContext<any>>,
  ) {
    this.jobIndexes = new Map(
      Array.from(indexes.entries()).map(([job, id]) => [id, job]),
    );

    for (const job of jobs) {
      for (const parent of Object.values(job.input ?? {})) {
        this.relations.push([parent, job]);
      }
    }
  }

  subscribeChange(cb: (event: Event) => void) {
    this.subs.add(cb);
    return () => this.subs.delete(cb);
  }

  private emitChange(event: Event) {
    this.subs.forEach((sub) => sub(event));
  }

  jobById(id: string) {
    const job = this.jobIndexes.get(id);
    if (!job) throw new Error(`Cannot found job with id ${id}`);
    return job;
  }

  indexOfJob(job: JobContext<any>) {
    const index = this.indexes.get(job);
    if (!index) throw new Error(`Cannot found index`, { cause: job });
    return index;
  }

  childrenOfJob(job: JobContext<any>) {
    return new Set(
      this.relations
        .filter(([parent, child]) => parent === job)
        .map(([, child]) => child),
    );
  }

  parentOfJob(job: JobContext<any>) {
    return new Set(
      this.relations
        .filter(([parent, child]) => job === child)
        .map(([parent]) => parent),
    );
  }

  stateOfJob(job: JobContext<any>) {
    return this.jobStates.get(job) ?? null;
  }

  async updateState(job: JobContext<any>, state: State): Promise<void> {
    this.jobStates.set(job, {
      ...this.jobStates.get(job),
      state,
    });

    this.emitChange({
      type: "updateState",
      timestamp: Date.now(),
      jobId: job.id,
      state: state,
    });
  }

  async resetState(job: JobContext<any>): Promise<void> {
    this.jobStates.delete(job);
  }

  async runJob(job: JobContext<any>) {
    const args: Record<string, unknown> = {};
    for (const [key, jobInput] of Object.entries(job.input ?? {})) {
      if (!(await this.storageService.has(jobInput.id)))
        throw new Error(`this job need run ${key}`);
      args[key] = await this.storageService.get(jobInput.id);
    }
    await this.updateState(job, State.Running);
    try {
      const res = await job.cb(args);
      await this.updateState(job, State.Success);
      await this.storageService.set(job.id, res);
      return res;
    } catch (ex) {
      await this.updateState(job, State.Failed);
      throw ex;
    }
  }

  *nextJobs(): Generator<JobContext<any>> {
    for (const job of this.jobs) {
      const parents = this.parentOfJob(job);
      if (
        this.stateOfJob(job)?.state !== State.Success &&
        Array.from(parents).every(
          (parent) => this.stateOfJob(parent)?.state === State.Success,
        )
      ) {
        yield job;
      }
    }
  }

  async *[Symbol.asyncIterator]() {
    let job: JobContext<any> | null = null;
    const next = () => this.nextJobs().next().value ?? null;
    while ((job = await next())) {
      yield job;
      await this.updateState(job, State.Success);
    }
  }

  toJSON(): RawJson {
    return {
      indexes: Array.from(this.indexes.values()),
      relations: this.relations.map(([a, b]) => [
        this.indexOfJob(a),
        this.indexOfJob(b),
      ]),
      jobStates: Object.fromEntries(
        Array.from(this.jobStates.entries()).map(([job, state]) => [
          this.indexOfJob(job),
          state,
        ]),
      ),
    };
  }

  static fromJson<J extends JobContext<any>>(raw: unknown, ...jobs: J[]) {
    if (!isRawJson(raw)) throw new Error(`invalid raw schema`, { cause: raw });

    const manager = this.create({ jobs: [...jobs] });
    const managerIndexes = Array.from(manager.indexes);
    if (managerIndexes.length !== raw.indexes.length)
      throw new Error(`${raw.indexes} jobs expected`);
    const indexesStr = new Set(manager.indexes.values());
    for (const indexJob of raw.indexes) {
      if (!indexesStr.has(indexJob)) {
        throw new Error(`Cannot found job id ${indexJob}`);
      }
    }

    const indexesJob = new Map(
      Array.from(manager.indexes.entries()).map(([job, jobIndex]) => [
        jobIndex,
        job,
      ]),
    );
    Object.entries(raw.jobStates).forEach(([jobIndex, jobState]) => {
      manager.jobStates.set(indexesJob.get(jobIndex)!, jobState);
    });

    return manager;
  }

  toFlowchart(options?: {
    formatNodeId?: (
      indexJob: string,
      job: JobContext<any>,
      manager: Manager,
    ) => string;
    formatLinkBetweenNodes?: (
      jobLeft: JobContext<any>,
      jobRight: JobContext<any>,
      manager: Manager,
    ) => string;
    formatNodeLabel?: (
      nodeId: string,
      job: JobContext<any>,
      manager: Manager,
    ) => string;
  }) {
    const formatNodeId =
      options?.formatNodeId ?? ((nodeId) => nodeId.replace(/\W/g, "_"));
    const formatLinkBetweenNodes =
      options?.formatLinkBetweenNodes ?? (() => `-->`);
    const formatNodeLabel =
      options?.formatNodeLabel ?? ((nodeId, job) => `${nodeId}`);

    return (
      `flowchart LR\n` +
      "\n" +
      Array.from(this.jobs)
        .map(
          (job) =>
            `${formatNodeLabel(formatNodeId(this.indexOfJob(job), job, this), job, this)}\n`,
        )
        .join("") +
      "\n" +
      Array.from(this.jobs)
        .map((job) => {
          return Array.from(this.childrenOfJob(job))
            .map((child) => {
              const idJobLeft: string = formatNodeId(
                this.indexOfJob(job),
                job,
                this,
              );
              const idJobRight: string = formatNodeId(
                this.indexOfJob(child),
                child,
                this,
              );
              const jobLeft: JobContext<any> = job;
              const jobRight: JobContext<any> = child;
              return `${idJobLeft} ${formatLinkBetweenNodes(jobLeft, jobRight, this)} ${idJobRight}\n`;
            })
            .join("");
        })
        .join("")
    );
  }

  static create<
    J extends JobContext<any>,
    O extends {
      /**
       * Session ID
       * @default {uuid value}
       */
      sid?: string;
      /**
       * Distinct ID
       */
      did?: string;
      dsn?: string;
      jobs: J[];
      signal?: AbortSignal;
    },
  >(
    options: O,
  ): O extends { signal: AbortSignal }
    ? Manager
    : Manager & { [Symbol.dispose]: () => void } {
    const sid = options.sid ?? crypto.randomUUID();
    const did = options.did;
    const jobs = options.jobs;
    const dsn = options.dsn;
    const defaultAbortControl = new AbortController();
    const signal = options.signal ?? defaultAbortControl.signal;

    let lastIndexGenerated = 0;
    const relationIndexes = new Map<JobContext<any>, string>();
    const jobsList = new Set<JobContext<any>>();
    const indexes = new Set<string>();

    const addJob = (job: JobContext<any>) => {
      if (relationIndexes.has(job)) return;

      const indexJob = job.id ?? `${++lastIndexGenerated}`;

      if (indexes.has(indexJob))
        throw new Error(`Already index ${indexJob} used`);
      indexes.add(indexJob);

      relationIndexes.set(job, indexJob);
      jobsList.add(job);

      Object.values(job.input ?? {}).forEach(addJob);
    };

    jobs.forEach(addJob);

    const manager = new Manager(sid, did, relationIndexes, jobsList);

    if (dsn) {
      DSN.create(manager, dsn, { signal });
    }

    if (signal === defaultAbortControl.signal) {
      Reflect.set(manager, Symbol.asyncDispose, async () => {
        const reason = new Closing();
        defaultAbortControl.abort(reason);
        await Promise.all(Array.from(reason.waitings));
      });
    }

    return manager as any;
  }

  static putStorageService(manager: Manager, storageService: StorageService) {
    manager.storageService = storageService;
  }

  static linkManager(managerFrom: Manager, managerTo: Manager) {
    managerFrom.subscribeChange((event) => {
      if (event.type === "updateState") {
        managerTo.updateState(managerTo.jobById(event.jobId), event.state);
      }
    });
  }
}
