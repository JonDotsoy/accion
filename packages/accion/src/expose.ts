import * as http from "http";
import { promiseWithResolvers } from "@accions/common/promise-with-resolvers";
import { debug } from "@accions/common/debug";
import { Manager } from "./manager.ts";
import * as handlers from "./expose-handlers/handlers.ts";

const ExposeDebug = debug("accion:expose");

const incomingMessageToUint8Array = async (
  incomingMessage: http.IncomingMessage,
) => {
  const { promise, resolve, reject } = promiseWithResolvers<Uint8Array>();

  let currentBuff: number[] = [];

  incomingMessage.on("data", (data: Uint8Array) => {
    currentBuff.push(...data);
  });

  incomingMessage.once("end", () => {
    resolve(new Uint8Array(currentBuff));
  });

  incomingMessage.once("error", (error) => {
    reject(error);
  });

  return promise;
};

class Server {
  public httpServer: http.Server;

  constructor(fetch: (request: Request) => Promise<Response>) {
    this.httpServer = http.createServer(async (req, res) => {
      try {
        const url = new URL(
          req.url ?? "",
          new URL(
            `http://${req.headers.host ?? "localhost"}`,
            "http://localhost",
          ),
        );
        const allowBody = (method: unknown) => {
          const m = `${method ?? ""}`.toLowerCase();
          if (m === "get") return false;
          if (m === "head") return false;
          return true;
        };

        const request = new Request(`${url}`, {
          headers: req.headers as any,
          method: req.method,
          body: allowBody(req.method)
            ? await incomingMessageToUint8Array(req)
            : undefined,
        });

        const response = await fetch(request);
        res.statusCode = response.status;

        for (const [k, v] of response.headers) {
          res.setHeader(k, v);
        }

        res.end(new Uint8Array(await response.arrayBuffer()));
      } catch (ex) {
        console.error(ex);
        res.statusCode = 500;
        res.end();
      }
    });
  }

  url() {
    if (this.httpServer) return serverToUrl(this.httpServer);
    return null;
  }

  async listen(port?: number, host?: string) {
    const { promise, resolve, reject } = promiseWithResolvers<void>();

    this.httpServer.addListener("error", reject);

    this.httpServer.listen(port, host, () => {
      this.httpServer.removeListener("error", reject);
      resolve();
    });

    return promise;
  }
}

const serverToUrl = (serve: http.Server) => {
  const rewriteHost = (host: string) => {
    if (host === "0.0.0.0") return `localhost`;
    return host;
  };
  const address = serve.address();
  if (typeof address === "string") return new URL(address, "http://localhost");
  if (address) {
    const url = new URL(`http://localhost`);
    url.host = rewriteHost(address.address);
    url.port = `${address.port}`;
    return url;
  }
  return address;
};

/**
 * Expose manager
 */
export class Expose {
  optionVerbose: boolean;
  debug = ExposeDebug;
  server: Server | null = null;

  constructor(
    private manager: Manager,
    options?: { verbose?: boolean },
  ) {
    this.optionVerbose = options?.verbose ?? false;

    if (this.optionVerbose) {
      this.debug = debug(ExposeDebug.namespace);
      this.debug.enabled = true;
    }
  }

  async close() {
    await new Promise<null>((resolve, reject) => {
      this.server?.httpServer.close((error) => {
        if (error) return reject(error);
        return resolve(null);
      });
    });
  }

  private async fetch(request: Request): Promise<Response> {
    const url = new URL(request.url);

    return (
      (await handlers.router(url, request, this.manager)) ??
      new Response("", { status: 404 })
    );
  }

  async listen(port: number = 6543, host: string = "127.0.0.1") {
    if (this.server !== null) throw new Error(`Server already listener`);

    const server = (this.server = new Server((req) => {
      let res: null | Response = null;
      return this.fetch(req)
        .then((r) => ((res = r), r))
        .finally(() => {
          this.debug(`${req.method} ${req.url} ${res?.status}`);
        });
    }));

    await server.listen(port, host);

    this.debug(`Server ready on ${this.server.url()}`);
  }
}
