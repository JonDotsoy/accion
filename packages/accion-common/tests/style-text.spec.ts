import { test, expect } from "bun:test";
import { styleText, Style } from "../src/style-text";

test("should style a style", () => {
    const demo = (styles: Style | Style[]) =>
        expect(styleText(styles as any, "text")).toMatchSnapshot(
            `text ${new Intl.ListFormat("en-us").format(Array.isArray(styles) ? styles : [styles])}`,
        );

    demo("underline");
    demo(["underline", "bold", "red"]);
    demo(["red"]);
});
