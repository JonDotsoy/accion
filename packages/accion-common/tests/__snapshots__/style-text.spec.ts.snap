// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`should style a style: text underline 1`] = `"[4mtext[24m"`;

exports[`should style a style: text underline, bold, and red 1`] = `"[31m[1m[4mtext[24m[22m[39m"`;

exports[`should style a style: text red 1`] = `"[31mtext[39m"`;
